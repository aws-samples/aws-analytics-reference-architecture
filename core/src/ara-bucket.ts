// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

import { IRole } from 'aws-cdk-lib/aws-iam';
import { IKey } from 'aws-cdk-lib/aws-kms';
import {
  BlockPublicAccess,
  Bucket,
  BucketAccessControl,
  BucketEncryption,
  BucketMetrics,
  CorsRule,
  IBucket,
  IntelligentTieringConfiguration,
  Inventory,
  LifecycleRule,
  ObjectOwnership,
} from 'aws-cdk-lib/aws-s3';
import { Aws, Duration, RemovalPolicy, Stack } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { SingletonKey } from './singleton-kms-key';

export interface AraBucketProps{
  /**
   * The Amazon S3 bucket name. The bucket name is postfixed with the AWS account ID and the AWS region
   */
  readonly bucketName: string;

  /**
   * The log file prefix to use for the bucket's access logs.
   * @default - access are not logged
   */
  readonly serverAccessLogsPrefix?: string;

  /**
   * Destination bucket for the server access logs.
   * @default - if serverAccessLogsPrefix is defined, use a unique bucket across the stack called `s3-access-logs`
   */
  readonly serverAccessLogsBucket?: IBucket;

  /**
   * The encryption mode for the bucket
   * @default - Server side encryption with AWS managed key (SSE-KMS)
   */
  readonly encryption?: BucketEncryption;

  /**
   * The KMS key for the bucket encryption
   * @default - if encryption is KMS, use a unique KMS key across the stack called `AraDefaultKmsKey`
   */
  readonly encryptionKey?: IKey ;

  /**
   * Enforces SSL for requests.
   * @default true
   */
  readonly enforceSSL?: boolean;

  /**
   * Specifies whether Amazon S3 should use an S3 Bucket Key with server-side encryption using KMS (SSE-KMS) for new objects in the bucket.
   * @default true
   */
  readonly bucketKeyEnabled?: boolean;

  /**
   * Policy to apply when the bucket is removed from this stack.
   * @default - destroy the bucket
   */
  readonly removalPolicy?: RemovalPolicy;

  /**
   * Whether all objects should be automatically deleted when the bucket is removed from the stack or when the stack is deleted.
   * Requires the `removalPolicy` to be set to `RemovalPolicy.DESTROY`.
   * @default true
   */
  readonly autoDeleteObjects?: boolean;

  /**
   * The block public access configuration of this bucket.
   * @default - Block all public access and no ACL or bucket policy can grant public access.
   */
  readonly blockPublicAccess?: BlockPublicAccess;

  /**
   * Whether this bucket should have versioning turned on or not.
   * @default false
   */
  readonly versioned?: boolean;

  /**
   * Rules that define how Amazon S3 manages objects during their lifetime.
   * @default - No lifecycle rules.
   */
  readonly lifecycleRules?: LifecycleRule[];

  /**
   * Specifies a canned ACL that grants predefined permissions to the bucket.
   * @default BucketAccessControl.PRIVATE
   */
  readonly accessControl?: BucketAccessControl;

  /**
   * Grants public read access to all objects in the bucket.
   * Similar to calling `bucket.grantPublicAccess()`
   * @default false
   */
  readonly publicReadAccess?: boolean;

  /**
   * The metrics configuration of this bucket.
   * @default - No metrics configuration.
   */
  readonly metrics?: BucketMetrics[];

  /**
   * The CORS configuration of this bucket.
   * @default - No CORS configuration.
   */
  readonly cors?: CorsRule[];

  /**
   * The inventory configuration of the bucket.
   * @default - No inventory configuration
  */
  readonly inventories?: Inventory[];

  /**
   * The objectOwnership of the bucket.
   * @default - No ObjectOwnership configuration, uploading account will own the object.
   */
  readonly objectOwnership?: ObjectOwnership;

  /**
   * Whether this bucket should have transfer acceleration turned on or not.
   * @default false
   */
  readonly transferAcceleration?: boolean;

  /**
   * The role to be used by the notifications handler
   * @default - a new role will be created.
   */
  readonly notificationsHandlerRole?: IRole;

  /**
   * Inteligent Tiering Configurations
   * @default No Intelligent Tiiering Configurations.
   */
  readonly intelligentTieringConfigurations?: IntelligentTieringConfiguration[];
}

/**
* An Amazon S3 Bucket following best practices for the AWS Analytics Reference Architecture.
* The bucket name is mandatory and is used as the CDK id.
* The bucket name is postfixed with the AWS account ID and the AWS region.
*
* The bucket has the following default properties:
*  * the encryption mode is KMS managed by AWS
*  * if the encryption mode is KMS customer managed, the encryption key is a default and unique KMS key for ARA
*  * the KMS key is used as a bucket key
*  * the SSL is enforced
*  * the objects are automatically deleted when the bucket is deleted
*  * the access are logged in a default and unique S3 bucket for ARA if serverAccessLogsPrefix is provided
*  * the access are not logged if serverAccessLogsPrefix is  not provided
*  * the public access is blocked and no bucket policy or object permission can grant public access
*
* All standard S3 Bucket properties can be provided to not use the defaults.
* Usage example:
* ```typescript
* import * as cdk from 'aws-cdk-lib';
* import { AraBucket } from 'aws-analytics-reference-architecture';
*
* const exampleApp = new cdk.App();
* const stack = new cdk.Stack(exampleApp, 'AraBucketStack');
*
* new AraBucket(stack, {
*  bucketName: 'test-bucket',
*  serverAccessLogsPrefix: 'test-bucket',
* });
* ```
*/
export class AraBucket extends Bucket {

  /**
  * Get the Amazon S3 Bucket from the AWS CDK Stack based on the provided name.
  * If no bucket exists, it creates a new one based on the provided properties.
  */
  public static getOrCreate(scope: Construct, props: AraBucketProps) {
    const stack = Stack.of(scope);
    const id = `${props.bucketName}`;

    const stackBucket = stack.nestedStackParent ? stack.nestedStackParent.node.tryFindChild(id) as Bucket : stack.node.tryFindChild(id) as Bucket;

    return stackBucket || new AraBucket(stack, props);
  }

  /**
  * Constructs a new instance of the AraBucket class
  * @param {Construct} scope the Scope of the CDK Construct
  * @param {AraBucketProps} props the AraBucketProps [properties]{@link AraBucketProps}
  * @access private
  */

  private constructor(scope: Construct, props: AraBucketProps) {

    var serverAccessLogsBucket = undefined;
    if ( props.serverAccessLogsPrefix ) {
      serverAccessLogsBucket = props.serverAccessLogsBucket || AraBucket.getOrCreate(scope, { bucketName: 's3-access-logs', encryption: BucketEncryption.S3_MANAGED });
    }

    // If using KMS encryption then use a customer managed key, if not set the key to undefined
    let bucketEncryptionKey: IKey | undefined = BucketEncryption.KMS == props.encryption ? props.encryptionKey || SingletonKey.getOrCreate(scope, 'DefaultKmsKey') : undefined;

    // If the bucket is for s3 access logs, we remove the bucketname to ensure uniqueness across stacks
    let bucketName = (props.bucketName == 's3-access-logs') ? undefined : `${props.bucketName}-${Aws.ACCOUNT_ID}-${Aws.REGION}`;
    // set the right default parameters in the S3 bucket props
    const bucketProps = {
      ...props,
      ...{
        bucketName: bucketName,
        encryption: props.encryption ? props.encryption : BucketEncryption.KMS_MANAGED,
        encryptionKey: bucketEncryptionKey,
        bucketKeyEnabled: BucketEncryption.KMS == props.encryption ? true : false,
        enforceSSL: props.enforceSSL || true,
        removalPolicy: props.removalPolicy || RemovalPolicy.DESTROY,
        autoDeleteObjects: props.autoDeleteObjects || true,
        serverAccessLogsBucket: serverAccessLogsBucket,
        serverAccessLogsPrefix: props.serverAccessLogsPrefix,
        blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
        lifecycleRules: props.lifecycleRules || [{ abortIncompleteMultipartUploadAfter: Duration.days(1) }],
      },
    };
    // build the S3 bucket
    super(scope, props.bucketName, bucketProps);
  }
}
