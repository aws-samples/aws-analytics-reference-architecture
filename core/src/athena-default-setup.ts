// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

import { CfnWorkGroup } from 'aws-cdk-lib/aws-athena';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';
import { SingletonBucket } from './singleton-bucket';

/**
 * AthenaDefaultSetup Construct to automatically setup a new Amazon Athena Workgroup with proper configuration for out-of-the-box usage
 */

export class AthenaDefaultSetup extends Construct {

  public readonly resultBucket: Bucket;

  /**
   * Constructs a new instance of the AthenaDefaultSetup class
   * @param {Construct} scope the Scope of the CDK Construct
   * @param {string} id the ID of the CDK Construct
   * @access public
   */

  constructor(scope: Construct, id: string) {
    super(scope, id);

    // Singleton Amazon S3 bucket for Amazon Athena Query logs
    this.resultBucket = SingletonBucket.getOrCreate(this, 'log');

    new CfnWorkGroup(this, 'athenaDefaultWorkgroup', {
      name: 'default',
      recursiveDeleteOption: true,
      workGroupConfiguration: {
        publishCloudWatchMetricsEnabled: false,
        resultConfiguration: {
          outputLocation: this.resultBucket.s3UrlForObject('athena-console-results'),
        },
      },
    });

  }
}