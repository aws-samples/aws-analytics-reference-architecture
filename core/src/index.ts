// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

export { ExampleProps, Example } from './example';
export { DataLakeStorageProps, DataLakeStorage } from './data-lake-storage';
export { SynchronousCrawlerProps, SynchronousCrawler } from './synchronous-crawler';
export { DatasetProps, Dataset, PartitionedDatasetProps, PartitionedDataset } from './datasets';
export { DataGeneratorProps, DataGenerator, BatchReplayerProps, BatchReplayer } from './data-generator';
export { SynchronousAthenaQueryProps, SynchronousAthenaQuery } from './synchronous-athena-query';
export { SingletonBucket } from './singleton-bucket';
export { Ec2SsmRole } from './ec2-ssm-role';
export { DataLakeCatalog } from './data-lake-catalog';
export { DataLakeExporter, DataLakeExporterProps } from './data-lake-exporter';
export { AthenaDefaultSetup } from './athena-default-setup';
export { SingletonGlueDefaultRole } from './singleton-glue-default-role';
export { EmrEksClusterProps, EmrEksCluster, EmrEksNodegroupOptions, EmrEksNodegroup, EmrVirtualClusterProps } from './emr-eks-data-platform';
export { FlywayRunner, FlywayRunnerProps } from './db-schema-manager';
