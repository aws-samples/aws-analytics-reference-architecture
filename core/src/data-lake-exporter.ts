import { Database, Table } from '@aws-cdk/aws-glue';
import { Role, ServicePrincipal, PolicyStatement } from '@aws-cdk/aws-iam';
import { Stream } from '@aws-cdk/aws-kinesis';
import { CfnDeliveryStream } from '@aws-cdk/aws-kinesisfirehose';
import { LogGroup, RetentionDays, LogStream } from '@aws-cdk/aws-logs';
import { Bucket, Location } from '@aws-cdk/aws-s3';
import { Construct, Aws, RemovalPolicy, Stack } from '@aws-cdk/core';


/**
 * The properties for DataLakeExporter Construct.
 */
export interface DataLakeExporterProps {
  /**
   * Sink must be an Amazon S3 Location composed of a bucket and a key.
   */
  readonly sinkLocation: Location;
  /**
   * Source must be an Amazon Kinesis Data Stream.
   */
  readonly sourceKinesisDataStream: Stream;
  /**
   * Source AWS Glue Database containing the schema of the stream.
   */
  readonly sourceGlueDatabase: Database;
  /**
   * Source AWS Glue Table containing the schema of the stream.
   */
  readonly sourceGlueTable: Table;
  /**
   * Delivery interval in seconds. The frequency of the data delivery is defined by this interval.
   * @default - Set to 900 seconds
   */
  readonly deliveryInterval?: Number;
  /**
   * Maximum delivery size in MB. The frequency of the data delivery is defined by this maximum delivery size.
   * @default - Set to 128 MB
   */
  readonly deliverySize?: Number;
}


/**
 * DataLakeExporter Construct to export data from a stream to the data lake.
 * Source can be an Amazon Kinesis Data Stream.
 * Target can be an Amazon S3 bucket.
 */
export class DataLakeExporter extends Construct {

  /**
   * Constructs a new instance of the DataLakeExporter class
   * @param {Construct} scope the Scope of the AWS CDK Construct
   * @param {string} id the ID of the AWS CDK Construct
   * @param {DataLakeExporterProps} props the DataLakeExporter [properties]{@link DataLakeExporterProps}
   * @access public
   */
  public readonly cfnIngestionStream: CfnDeliveryStream;

  constructor(scope: Construct, id: string, props: DataLakeExporterProps) {
    super(scope, id);

    if ( props.deliverySize || 128 > 128 ) { throw 'deliverySize cannot be more than 128MB';}
    if ( props.deliveryInterval || 900 > 900 ) { throw 'deliveryInterval cannot be more than 900s';}

    const stack = Stack.of(this);

    // Get the Bucket from Amazon S3 Location sink
    const sinkBucket = Bucket.fromBucketName(this, 'sinkBucket', props.sinkLocation.bucketName);

    // Create log group for storing Amazon Kinesis Firehose logs.
    const logGroup = new LogGroup(this, 'dataLakeExporterLogGroup', {
      logGroupName: '/data-lake-exporter/',
      removalPolicy: RemovalPolicy.DESTROY,
      retention: RetentionDays.ONE_DAY,
    });

    // Create the Kinesis Firehose log stream.
    const firehoseLogStream = new LogStream(this, 'dataLakeExporterLogStream', {
      logGroup: logGroup,
      logStreamName: 'firehose-stream',
      removalPolicy: RemovalPolicy.DESTROY,
    });

    // Create an Amazon IAM Role used by Amazon Kinesis Firehose delivery stream
    const role = new Role(this, 'dataLakeExporterRole', {
      assumedBy: new ServicePrincipal('firehose.amazonaws.com'),
    });

    // add policy to access AWS Glue Database and Table for format conversion
    role.addToPolicy(
      new PolicyStatement({
        resources: [
          props.sourceGlueTable.tableArn,
          props.sourceGlueDatabase.catalogArn,
          props.sourceGlueDatabase.databaseArn,
        ],
        actions: [
          'glue:GetTable',
          'glue:GetTableVersion',
          'glue:GetTableVersions',
        ],
      }),
    );

    // add policy to access source stream
    role.addToPolicy(
      new PolicyStatement({
        resources: [
          props.sourceKinesisDataStream.streamArn,
        ],
        actions: [
          'kinesis:DescribeStream',
          'kinesis:GetShardIterator',
          'kinesis:GetRecords',
          'kinesis:ListShards',
        ],
      }),
    );

    // add policy to write data in Amazon S3 Location
    role.addToPolicy(
      new PolicyStatement({
        resources: [
          stack.formatArn({
            account: '',
            region: '',
            service: 's3',
            resource: props.sinkLocation.bucketName,
            resourceName: props.sinkLocation.objectKey,
          }),
          stack.formatArn({
            account: '',
            region: '',
            service: 's3',
            resource: props.sinkLocation.bucketName,
            resourceName: `${props.sinkLocation.objectKey}/*`,
          }),
          stack.formatArn({
            account: '',
            region: '',
            service: 's3',
            resource: props.sinkLocation.bucketName,
          }),
        ],
        actions: [
          's3:AbortMultipartUpload',
          's3:GetBucketLocation',
          's3:GetObject',
          's3:ListBucket',
          's3:ListBucketMultipartUploads',
          's3:PutObject',
        ],
      }),
    );

    // add policy to write logs to CloudWatch logs
    role.addToPolicy(
      new PolicyStatement({
        resources: [
          `${logGroup.logGroupArn}:log-stream:${firehoseLogStream.logStreamName}`,
        ],
        actions: [
          'logs:PutLogEvents',
        ],
      }),
    );

    // TODO add policy for KMS managed?

    /*     this.ingestionStream = new DeliveryStream(this, 'dataLakeExporter', {
      sourceStream: props.sourceKinesisDataStream,
      destinations: [new S3Bucket(sinkBucket,{
        dataOutputPrefix: props.sinkLocation.objectKey,
        errorOutputPrefix: `${props.sinkLocation.objectKey}-error`,
        logGroup: logGroup,
        compression: Compression.SNAPPY,
        bufferingInterval: Duration.seconds(props.deliveryInterval || 900),
        bufferingSize: Size.mebibytes(props.deliverySize || 128),
      })],
      encryption: StreamEncryption.AWS_OWNED,
    }) */

    // Create the Delivery stream from Cfn because L2 Construct doesn't support conversion to parquet and custom partitioning
    this.cfnIngestionStream = new CfnDeliveryStream(this, 'dataLakeExporter', {
      deliveryStreamType: 'KinesisStreamAsSource',
      deliveryStreamEncryptionConfigurationInput: {
        keyType: 'AWS_OWNED_CMK',
      },
      extendedS3DestinationConfiguration: {
        bucketArn: sinkBucket.bucketArn,
        bufferingHints: {
          intervalInSeconds: props.deliveryInterval || 900,
          sizeInMBs: props.deliverySize || 128,
        },
        cloudWatchLoggingOptions: {
          logGroupName: logGroup.logGroupName,
          logStreamName: firehoseLogStream.logStreamName,
        },
        roleArn: role.roleArn,
        errorOutputPrefix: `${props.sinkLocation.objectKey}-error`,
        prefix: props.sinkLocation.objectKey,
        compressionFormat: 'UNCOMPRESSED',
        s3BackupMode: 'Disabled',
        dataFormatConversionConfiguration: {
          enabled: true,
          inputFormatConfiguration: {
            deserializer: {
              openXJsonSerDe: {},
            },
          },
          outputFormatConfiguration: {
            serializer: {
              parquetSerDe: {},
            },
          },
          schemaConfiguration: {
            roleArn: role.roleArn,
            catalogId: Aws.ACCOUNT_ID,
            region: Aws.REGION,
            databaseName: props.sourceGlueDatabase.databaseName,
            tableName: props.sourceGlueTable.tableName,
          },
        },
      },
      kinesisStreamSourceConfiguration: {
        kinesisStreamArn: '',
        roleArn: role.roleArn,
      },
    });
  }
}
