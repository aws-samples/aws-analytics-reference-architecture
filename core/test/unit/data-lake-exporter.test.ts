// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

/**
 * Tests Datalake Exporter
 *
 * @group unit/datalake/exporter
 */

import { RetentionDays } from 'aws-cdk-lib/aws-logs';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { Stack } from 'aws-cdk-lib';
import { DataLakeExporter } from '../../src/data-lake-exporter';
import { Database, DataFormat, Table } from '@aws-cdk/aws-glue-alpha';
import { Template } from 'aws-cdk-lib/assertions';
import { Stream } from 'aws-cdk-lib/aws-kinesis';


test('dataLakeExporter', () => {
  const dataLakeExporterStack = new Stack();

  const stream = new Stream(dataLakeExporterStack, 'testStream');
  const db = new Database(dataLakeExporterStack, 'testDB', { databaseName: 'test_db' });
  const table = new Table(dataLakeExporterStack, 'testTable', {
    database: db,
    tableName: 'test_table',
    dataFormat: DataFormat.JSON,
    columns: [
      {
        name: 'a',
        type: {
          isPrimitive: true,
          inputString: 'STRING',
        },
      },
    ],
  });
  const bucket = new Bucket(dataLakeExporterStack, 'Bucket');

  new DataLakeExporter(dataLakeExporterStack, 'testExporter', {
    sinkBucket: bucket,
    sinkObjectKey: 'test',
    sourceKinesisDataStream: stream,
    sourceGlueDatabase: db,
    sourceGlueTable: table,
  });

  const template = Template.fromStack(dataLakeExporterStack);

  // TODO: add tests (refer to data-lake-storage.test.ts)
  // Test if the stack has S3 Buckets
  template.hasResource('AWS::S3::Bucket', 1);

  // Test if stack has log group
  template.hasResourceProperties('AWS::Logs::LogGroup', {
    LogGroupName: '/aws/data-lake-exporter/',
    RetentionInDays: RetentionDays.ONE_WEEK,
  });

  //Test stack for firehose stream
  template.hasResourceProperties('AWS::KinesisFirehose::DeliveryStream', {
    ExtendedS3DestinationConfiguration: {
      BufferingHints: {
        IntervalInSeconds: 900,
        SizeInMBs: 128,
      },
    },
  });

  expect(template).toMatchSnapshot();
});
