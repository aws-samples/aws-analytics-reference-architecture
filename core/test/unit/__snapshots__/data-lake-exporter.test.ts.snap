// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dataLakeExporter 1`] = `
Object {
  "Conditions": Object {
    "AwsCdkKinesisEncryptedStreamsUnsupportedRegions": Object {
      "Fn::Or": Array [
        Object {
          "Fn::Equals": Array [
            Object {
              "Ref": "AWS::Region",
            },
            "cn-north-1",
          ],
        },
        Object {
          "Fn::Equals": Array [
            Object {
              "Ref": "AWS::Region",
            },
            "cn-northwest-1",
          ],
        },
      ],
    },
  },
  "Description": "undefined (uksb-1sljh1bja)",
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "Bucket83908E77": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "testDB74E9B281": Object {
      "Properties": Object {
        "CatalogId": Object {
          "Ref": "AWS::AccountId",
        },
        "DatabaseInput": Object {
          "Name": "test_db",
        },
      },
      "Type": "AWS::Glue::Database",
    },
    "testExporterdataLakeExporter51BD8EA9": Object {
      "DependsOn": Array [
        "testExporterdataLakeExporterRoleDefaultPolicy3EEE6D56",
      ],
      "Properties": Object {
        "DeliveryStreamType": "KinesisStreamAsSource",
        "ExtendedS3DestinationConfiguration": Object {
          "BucketARN": Object {
            "Fn::GetAtt": Array [
              "Bucket83908E77",
              "Arn",
            ],
          },
          "BufferingHints": Object {
            "IntervalInSeconds": 900,
            "SizeInMBs": 128,
          },
          "CloudWatchLoggingOptions": Object {
            "LogGroupName": Object {
              "Ref": "testExporterdataLakeExporterLogGroup610F71C7",
            },
            "LogStreamName": Object {
              "Ref": "testExporterdataLakeExporterLogStreamF2BE8809",
            },
          },
          "CompressionFormat": "UNCOMPRESSED",
          "DataFormatConversionConfiguration": Object {
            "Enabled": true,
            "InputFormatConfiguration": Object {
              "Deserializer": Object {
                "OpenXJsonSerDe": Object {},
              },
            },
            "OutputFormatConfiguration": Object {
              "Serializer": Object {
                "ParquetSerDe": Object {},
              },
            },
            "SchemaConfiguration": Object {
              "CatalogId": Object {
                "Ref": "AWS::AccountId",
              },
              "DatabaseName": Object {
                "Ref": "testDB74E9B281",
              },
              "Region": Object {
                "Ref": "AWS::Region",
              },
              "RoleARN": Object {
                "Fn::GetAtt": Array [
                  "testExporterdataLakeExporterRole1BEE617B",
                  "Arn",
                ],
              },
              "TableName": Object {
                "Ref": "testTable12C06BD8",
              },
            },
          },
          "ErrorOutputPrefix": "test-error",
          "Prefix": "test",
          "RoleARN": Object {
            "Fn::GetAtt": Array [
              "testExporterdataLakeExporterRole1BEE617B",
              "Arn",
            ],
          },
          "S3BackupMode": "Disabled",
        },
        "KinesisStreamSourceConfiguration": Object {
          "KinesisStreamARN": Object {
            "Fn::GetAtt": Array [
              "testStream8BCA7523",
              "Arn",
            ],
          },
          "RoleARN": Object {
            "Fn::GetAtt": Array [
              "testExporterdataLakeExporterRole1BEE617B",
              "Arn",
            ],
          },
        },
      },
      "Type": "AWS::KinesisFirehose::DeliveryStream",
    },
    "testExporterdataLakeExporterLogGroup610F71C7": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "LogGroupName": "/aws/data-lake-exporter/",
        "RetentionInDays": 7,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "testExporterdataLakeExporterLogStreamF2BE8809": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "LogGroupName": Object {
          "Ref": "testExporterdataLakeExporterLogGroup610F71C7",
        },
        "LogStreamName": "firehose-stream",
      },
      "Type": "AWS::Logs::LogStream",
      "UpdateReplacePolicy": "Delete",
    },
    "testExporterdataLakeExporterRole1BEE617B": Object {
      "DependsOn": Array [
        "testExportermanagedPolicyKinesisFirehoseE66FDFBA",
      ],
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "firehose.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Ref": "testExportermanagedPolicyKinesisFirehoseE66FDFBA",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "testExporterdataLakeExporterRoleDefaultPolicy3EEE6D56": Object {
      "DependsOn": Array [
        "testExportermanagedPolicyKinesisFirehoseE66FDFBA",
      ],
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "Bucket83908E77",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "Bucket83908E77",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "kinesis:DescribeStreamSummary",
                "kinesis:GetRecords",
                "kinesis:GetShardIterator",
                "kinesis:ListShards",
                "kinesis:SubscribeToShard",
                "kinesis:DescribeStream",
                "kinesis:ListStreams",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "testStream8BCA7523",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": Array [
                "glue:BatchGetPartition",
                "glue:GetPartition",
                "glue:GetPartitions",
                "glue:GetTable",
                "glue:GetTables",
                "glue:GetTableVersion",
                "glue:GetTableVersions",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":glue:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":table/",
                    Object {
                      "Ref": "testDB74E9B281",
                    },
                    "/",
                    Object {
                      "Ref": "testTable12C06BD8",
                    },
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "testTableBucket030117E6",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "testTableBucket030117E6",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "glue:GetTableVersions",
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":glue:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":table/",
                      Object {
                        "Ref": "testDB74E9B281",
                      },
                      "/",
                      Object {
                        "Ref": "testTable12C06BD8",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":glue:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":catalog",
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":glue:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":database/",
                      Object {
                        "Ref": "testDB74E9B281",
                      },
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "testExporterdataLakeExporterRoleDefaultPolicy3EEE6D56",
        "Roles": Array [
          Object {
            "Ref": "testExporterdataLakeExporterRole1BEE617B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "testExportermanagedPolicyKinesisFirehoseE66FDFBA": Object {
      "Properties": Object {
        "Description": "",
        "Path": "/",
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "logs:PutLogEvents",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "testExporterdataLakeExporterLogGroup610F71C7",
                        "Arn",
                      ],
                    },
                    ":log-stream:",
                    Object {
                      "Ref": "testExporterdataLakeExporterLogStreamF2BE8809",
                    },
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::ManagedPolicy",
    },
    "testStream8BCA7523": Object {
      "Properties": Object {
        "RetentionPeriodHours": 24,
        "ShardCount": 1,
        "StreamEncryption": Object {
          "Fn::If": Array [
            "AwsCdkKinesisEncryptedStreamsUnsupportedRegions",
            Object {
              "Ref": "AWS::NoValue",
            },
            Object {
              "EncryptionType": "KMS",
              "KeyId": "alias/aws/kinesis",
            },
          ],
        },
        "StreamModeDetails": Object {
          "StreamMode": "PROVISIONED",
        },
      },
      "Type": "AWS::Kinesis::Stream",
    },
    "testTable12C06BD8": Object {
      "Properties": Object {
        "CatalogId": Object {
          "Ref": "AWS::AccountId",
        },
        "DatabaseName": Object {
          "Ref": "testDB74E9B281",
        },
        "TableInput": Object {
          "Description": "test_table generated by CDK",
          "Name": "test_table",
          "Parameters": Object {
            "classification": "json",
            "has_encrypted_data": false,
          },
          "StorageDescriptor": Object {
            "Columns": Array [
              Object {
                "Name": "a",
                "Type": "STRING",
              },
            ],
            "Compressed": false,
            "InputFormat": "org.apache.hadoop.mapred.TextInputFormat",
            "Location": Object {
              "Fn::Join": Array [
                "",
                Array [
                  "s3://",
                  Object {
                    "Ref": "testTableBucket030117E6",
                  },
                  "/",
                ],
              ],
            },
            "OutputFormat": "org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat",
            "SerdeInfo": Object {
              "SerializationLibrary": "org.openx.data.jsonserde.JsonSerDe",
            },
            "StoredAsSubDirectories": false,
          },
          "TableType": "EXTERNAL_TABLE",
        },
      },
      "Type": "AWS::Glue::Table",
    },
    "testTableBucket030117E6": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
